lib/puppet/network/http/connection.rb

```diff
@@ -198,6 +198,7 @@ def do_request(request, options)
               current_request[header] = value
             end
           when 429, 503
+            connection.finish
```

Comment:
Seems reasonable to me. Should we add a `if http.started?` guard here just in case the connection is already closed? For example, the `Connection: close` header added to Puppet Server might cause the connection to be closed by the time we get to this logic on the client side.